<h1 align="center"> IST 5520 Milestone 2: Data Analysis I <h1>
<h3 align="center"> AirBnB Dataset <h3>

#### Student: Ronald Adomako
#### Student: Idris Dobbs
#### Student: Narendra Chigili
#### Student: Nikhil Srirama Sai

# Instruction 1: Cleanse and visualize data

## Introduction

AirBnB is a PaaS for the short term rental market. Users use the platform to list residences for short term rentals. We noticed that for big cities, such as New York City, there are many host and some may want to use AirBnB for lucrative means. Given location and characteristics of a property, a new host would want to know whether he or she is charging the optimal amount to rent the space to lodgers.

We noticed that from the New York City dataset for AirBnB, the categorical variable for location  was too coarse. The descriptor says the location is categorized into five boroughs. For a big city such as New York City, there are a lot of insights missing from a business perspective because neighborhoods vary drastically in property amenities even within a single borough. Moreso, the geo-coordinates are too fine for business purposes. To handle this we implement a zip code converter to categorize properties based on their location - Feature Selection.

- What are the largest determinants / predictors of AirBnB rental prices?
- How can we optimize rental revenue based on rental location and other characteristics?
- What price should be charged based on rental location/ characteristics?

We are opperating under the assumption that NYC AirBnB prices has reached a *steady-state*: i.e. the market has been active for long enough in NYC and there are enough data points (observations) in NYC that the **mean** is meaningful.

We want to know whether a host is charging an optimal price. To do this we group the observations by neigborhood and then take the average price. Hosts who charge at or above this price are considered optimal in their respective neighborhood while hosts who charge below the average price for their neighborhood are sub-optimal. Consider the case where all hosts charge the same price within a neighborhood, then the mean is the mode is the median- uniform data, no variance. All the hosts in this neighborhood would be optimal.

Consider the case where one of those hosts charge below what would have been the average, then only that host is sub-optimal while the rest are optimal. Conversely, if one host charge above the rest while everyone else charges the same, then that one host would be optimal while the rest are suboptimal.

Along with the *steady-state* assumption, by grouping the data by neighborhoods we assume that on average homes and amenities are similar by neighborhood. The geo-coordinates are too fine a scale and the boroughs are too coarse a scale. A meso-scale would be by zipcode, which we would expect to have higher precision of similarities between host, or by neighborhood. For a dataset with 39881 observations, transforming geo-coordinates prove to be computationally expensive (22 hours on standard household computer). We chose the next best meso-scale: what AirBnB features as "neighbourhood".

From a business perspective, we want to know what percent of hosts per neighborhood are charging an optimal price, and aggregating this data the percent hosts charge an optimal price in NYC overall. We see that using the neighborhood grouping allows us to compare on a common scale for all hosts. We don't have hosts income, so we wouldn't be able to measure profit. Likewise, revenue wouldn't be a fair scale because hosts with more units will outperform host with smaller units just by volume. A nieghborhood comparison allows a better metric to assess price per room, where we expect reasonably small variance per neighborhood. Furthermore, comparing by percentage is normalizes our comparison in general.

## Data Source and Collection

#### We chose the AirBnB dataset for New York City (NYC). We want to build a model that indicates whether hosts are charging an optimum amount for their rental.

http://insideairbnb.com/new-york-city/

[data dictionary](https://docs.google.com/spreadsheets/d/1iWCNJcSutYqpULSQHlNyGInUvHg2BoUGoNRIGa6Szc4/edit?usp=sharing)

http://insideairbnb.com/get-the-data

http://data.insideairbnb.com/united-states/ny/new-york-city/2022-09-07/visualisations/listings.csv



The data dictionary for New York City AirBnB dataset consists of 75 columns (variables or features) and 39881 observations. Based on our research question and several [Kaggle challenges](https://www.kaggle.com/search?q=airbnb-listing-in-nyc) (www.kaggle.com/search?q=airbnb-listing-in-nyc) we chose the following features for preliminary analysis.

In [34]:
import pandas as pd
import matplotlib.pyplot as plt
import sidetable as stb
import numpy as np
import seaborn as sns

%matplotlib inline

In [1]:

csv_URL = "http://data.insideairbnb.com/united-states/ny/new-york-city/2022-09-07/visualisations/listings.csv"
df2 = pd.read_csv(csv_URL)

In [2]:
pd.DataFrame(df2.columns, columns=['Features'])

Features
0                               id
1                             name
2                          host_id
3                        host_name
4              neighbourhood_group
5                    neighbourhood
6                         latitude
7                        longitude
8                        room_type
9                            price
10                  minimum_nights
11               number_of_reviews
12                     last_review
13               reviews_per_month
14  calculated_host_listings_count
15                availability_365
16           number_of_reviews_ltm
17                         license

The full data dictionary has the following descriptions:

In [3]:
data = pd.read_csv('data_dictionary.csv')
pd.set_option('display.max_rows',1000)
data

Field                       Type  \
0                                             id                    integer   
1                                    listing_url                       text   
2                                      scrape_id                     bigint   
3                                   last_scraped                   datetime   
4                                         source                       text   
5                                           name                       text   
6                                    description                       text   
7                          neighborhood_overview                       text   
8                                    picture_url                       text   
9                                        host_id                    integer   
10                                      host_url                       text   
11                                     host_name                       text   
12                                    host_since                       date   
13                                 host_location                       text   
14                                    host_about                       text   
15                            host_response_time                        NaN   
16                            host_response_rate                        NaN   
17                          host_acceptance_rate                        NaN   
18                             host_is_superhost  boolean [t=true; f=false]   
19                            host_thumbnail_url                       text   
20                              host_picture_url                       text   
21                            host_neighbourhood                       text   
22                           host_listings_count                       text   
23                     host_total_listings_count                       text   
24                            host_verifications                        NaN   
25                          host_has_profile_pic  boolean [t=true; f=false]   
26                        host_identity_verified  boolean [t=true; f=false]   
27                                 neighbourhood                       text   
28                        neighbourhood_cleansed                       text   
29                  neighbourhood_group_cleansed                       text   
30                                      latitude                    numeric   
31                                     longitude                    numeric   
32                                 property_type                       text   
33                                     room_type                       text   
34                                           NaN                        NaN   
35                                           NaN                        NaN   
36                                           NaN                        NaN   
37                                           NaN                        NaN   
38                                           NaN                        NaN   
39                                           NaN                        NaN   
40                                           NaN                        NaN   
41                                           NaN                        NaN   
42                                           NaN                        NaN   
43                                           NaN                        NaN   
44                                           NaN                        NaN   
45                                           NaN                        NaN   
46                                           NaN                        NaN   
47                                           NaN                        NaN   
48                                  accommodates                    integer   
49                                     bathrooms                    numeric   
50            

In [4]:
df2

id                                               name  \
0                   77765                               Superior @ Box House   
1                    2539                 Clean & quiet apt home by the park   
2                   45910                 Beautiful Queens Brownstone! - 5BR   
3                   45935                       Room in Beautiful Townhouse.   
4                   45936                 Couldn't Be Closer To Columbia Uni   
...                   ...                                                ...   
39876            27577588           Luxury Studio ON Grove Street E0C - B1CA   
39877  654151117629853651                        Lovely 3- bedroom apartment   
39878  553754115911961053          Trendy 3-bedroom apartment near Manhattan   
39879  698195550745703156  Luxurious private waterfront terrace, 2BR 2BA Apt   
39880            48971505           Just Blocks to Grove PATH and JC Med Ctr   

         host_id            host_name neighbourhood_group  \
0         417504  The Box House Hotel            Brooklyn   
1           2787                 John            Brooklyn   
2         204539                 Mark              Queens   
3         204586                    L               Bronx   
4         867225                Rahul           Manhattan   
...          ...                  ...                 ...   
39876   37412692                  Kim           Manhattan   
39877  117540494               Miriam              Queens   
39878   15048320                India           Manhattan   
39879  151487807                Asser            Brooklyn   
39880      46201                    J           Manhattan   

             neighbourhood   latitude  longitude        room_type  price  \
0               Greenpoint  40.737770 -73.953660       Hotel room    308   
1               Kensington  40.645290 -73.972380     Private room    299   
2                Ridgewood  40.703090 -73.899630  Entire home/apt    425   
3               Mott Haven  40.806350 -73.922010     Private room     60   
4      Morningside Heights  40.806300 -73.959850     Private room     75   
...                    ...        ...        ...              ...    ...   
39876         Ellis Island  40.718220 -74.037940  Entire home/apt    135   
39877             Rosedale  40.647244 -73.720088  Entire home/apt    180   
39878      Upper West Side  40.787320 -74.004470  Entire home/apt    240   
39879         Williamsburg  40.709192 -73.970121  Entire home/apt    400   
39880         Ellis Island  40.718350 -74.044160     Private room     40   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
0                   2                 42  2022-07-18               0.30   
1                  30                  9  2018-10-19               0.11   
2                  30                 13  2019-11-12               0.10   
3                  30                  0         NaN                NaN   
4                  31                135  2022-07-11               0.95   
...               ...                ...         ...                ...   
39876             365                  2  2019-09-16               0.04   
39877               1                  5  2022-08-24               1.92   
39878               5                 18  2022-08-22               2.87   
39879              30                  0         NaN                NaN   
39880               1                 15  2021-10-25               0.91   

       calculated_host_listings_count  availability_365  \
0                                  30               217   
1                                   9               356   
2                                   6               365   
3                                   1                83   
4                                   1               219   
...                               ...               ...   
39876                               7               365   
39877                               1      

## Data Manipulation

In [33]:
#df2.groupby?

In [31]:
#Create a data frame grouping by neighborhood for average price
hood_price_obj = df2[['neighbourhood','price']].groupby('neighbourhood')
df_mean_price = hood_price_obj.mean()
df_mean_price[['price']] = df_mean_price[['price']].round(2)
df_mean_price

price
neighbourhood                       
Allerton                      118.78
Arden Heights                 113.86
Arrochar                      132.06
Arverne                       230.26
Astoria                       109.01
Bath Beach                    117.19
Battery Park City             230.90
Bay Ridge                     121.55
Bay Terrace                   151.25
Baychester                    111.28
Bayside                       182.70
Bayswater                     106.75
Bedford-Stuyvesant            140.85
Belle Harbor                  354.86
Bellerose                     132.82
Belmont                       113.73
Bensonhurst                    96.74
Bergen Beach                  281.17
Bloomfield                    188.00
Boerum Hill                   249.72
Borough Park                   86.93
Breezy Point                  173.00
Briarwood                     390.86
Brighton Beach                109.43
Broad Channel                 254.00
Bronx Park                    123.60
Bronxdale                      82.11
Brooklyn Heights              243.96
Brownsville                   149.77
Bull's Head                    60.00
Bushwick                      114.35
Cambria Heights               137.44
Canarsie                      166.15
Carroll Gardens               228.87
Castle Hill                   110.75
Castleton Corners             115.86
Central Park                  298.08
Chelsea                       275.72
Chelsea, Staten Island         90.00
Chinatown                     240.10
City Island                   168.60
Civic Center                  237.57
Claremont Village             106.82
Clason Point                  110.67
Clifton                       165.73
Clinton Hill                  198.50
Co-op City                    110.00
Cobble Hill                   250.74
College Point                 114.75
Columbia St                   200.74
Concord                       113.12
Concourse                     109.56
Concourse Village              86.68
Coney Island                  131.68
Corona                        100.26
Country Club                   98.00
Crotona Park                  104.00
Crown Heights                 143.16
Cypress Hills                 112.88
DUMBO                         281.08
Ditmars Steinway              122.32
Dongan Hills                  155.17
Douglaston                     89.86
Downtown Brooklyn             248.22
Dyker Heights                 105.24
East Elmhurst                 140.92
East Flatbush                 132.93
East Harlem                   193.00
East Morrisania               107.56
East New York                 121.00
East Village                  221.96
Eastchester                   134.21
Edenwald                      105.20
Edgemere                      179.93
Ellis Island                  198.28
Elmhurst                       93.11
Eltingville                   210.80
Emerson Hill                  111.75
Far Rockaway                  230.82
Ferry Point Park              225.00
Fieldston                      88.70
Financial District            347.61
Flatbush                      150.74
Flatiron District             407.53
Flatlands                     144.50
Floral Park                   109.27
Flushing                      105.47
Flushing Meadows Corona Park  101.80
Fordham                        87.17
Forest Hills                  179.52
Forest Park                   100.50
Fort Greene                   194.05
Fort Hamilton                 106.11
Fort Wadsworth                650.00
Fresh Meadows                 137.31
Freshkills Park               141.50
Gerritsen Beach               225.80
Glendale                      131.16
Gowanus                       239.26
Gramercy                      259.63
Graniteville                   83.75
Grant City                     60.14
Gravesend                     127.24
Great Kills                   129.75
Green-Wood Cemetery           160.90
Greenpoint                    198.15
Greenwich Village             338.0

In [30]:
df_mean_price.info()

<class 'pandas.core.frame.DataFrame'>
Index: 244 entries, Allerton to Woodside
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   244 non-null    float64
dtypes: float64(1)
memory usage: 3.8+ KB


**We have reduced 39881 into 244 rows of manageable data!**

In [80]:
df2.stb.freq(['neighbourhood'])

neighbourhood  count   percent  cumulative_count  \
0              Bedford-Stuyvesant   2779  6.968230              2779   
1                    Williamsburg   2456  6.158321              5235   
2                          Harlem   1878  4.709009              7113   
3                         Midtown   1701  4.265189              8814   
4                        Bushwick   1657  4.154861             10471   
5                 Upper West Side   1630  4.087159             12101   
6                  Hell's Kitchen   1578  3.956771             13679   
7                 Upper East Side   1379  3.457787             15058   
8                   Crown Heights   1197  3.001429             16255   
9                    East Village   1057  2.650385             17312   
10                        Chelsea    901  2.259221             18213   
11                Lower East Side    797  1.998445             19010   
12                    East Harlem    768  1.925729             19778   
13                     Greenpoint    708  1.775281             20486   
14                        Astoria    686  1.720117             21172   
15             Washington Heights    587  1.471879             21759   
16                   West Village    572  1.434267             22331   
17                  East Flatbush    530  1.328954             22861   
18                       Flatbush    488  1.223640             23349   
19                       Flushing    485  1.216118             23834   
20             Financial District    482  1.208596             24316   
21               Long Island City    477  1.196058             24793   
22      Prospect-Lefferts Gardens    430  1.078208             25223   
23                       Kips Bay    408  1.023044             25631   
24                    Murray Hill    371  0.930268             26002   
25                   Clinton Hill    367  0.920238             26369   
26                       Woodside    336  0.842506             26705   
27                  East New York    326  0.817432             27031   
28                     Park Slope    324  0.812417             27355   
29               Theater District    322  0.807402             27677   
30                    Fort Greene    322  0.807402             27999   
31                      Ridgewood    315  0.789850             28314   
32                        Jamaica    312  0.782327             28626   
33                    Sunset Park    297  0.744716             28923   
34                      Sunnyside    284  0.712119             29207   
35               Ditmars Steinway    253  0.634387             29460   
36                       Elmhurst    248  0.621850             29708   
37                        Tribeca    246  0.616835             29954   
38                           SoHo    242  0.606805             30196   
39                      Chinatown    237  0.594268             30433   
40                  East Elmhurst    234  0.586746             30667   
41                       Gramercy    227  0.569193             30894   
42                         Nolita    223  0.559164             31117   
43               Prospect Heights    217  0.544119             31334   
44                       Canarsie    211  0.529074             31545   
45                Jackson Heights    207  0.519044             31752   
46              Greenwich Village    206  0.516537             31958   
47                      Wakefield    192  0.481432             32150   
48                    South Slope    190  0.476417             32340   
49            Morningside Heights    188  0.471402             32528   
50                         Inwood    186  0.466388             32714   
51                Carroll Gardens    179  0.448835             32893   
52                 Sheepshead Bay    172  0.431283             33065   
53                  Cypress Hills    168  0.421253             33233   
54                        Gowanus    166  0.416238          

In [82]:
#df2.stb.freq(['neighbourhood']).describe()

In [50]:

#df = df2.sort_values('neighbourhood')
df_hood = df2.stb.freq(['neighbourhood'])
df_hood = df_hood.loc[:,'neighbourhood':'percent']
df_hood = df_hood.sort_values('neighbourhood')


In [55]:
df_hood.reset_index(inplace=True)


In [71]:
df_mean = df_mean_price[['price']].reset_index()
df_mean.head()

neighbourhood   price
0       Allerton  118.78
1  Arden Heights  113.86
2       Arrochar  132.06
3        Arverne  230.26
4        Astoria  109.01

In [73]:
df_hood['price'] = df_mean['price']
df_hood.rename(columns={'index':'pop_rank'}, inplace=True)
df_hood.head()

pop_rank  neighbourhood  count   percent   price
0       107       Allerton     45  0.112836  118.78
1       204  Arden Heights      7  0.017552  113.86
2       158       Arrochar     17  0.042627  132.06
3        64        Arverne    110  0.275821  230.26
4        14        Astoria    686  1.720117  109.01

In [77]:
df_hood['opt_percent'] = /df_hood['percent']
df_hood.head()

ValueError: Can only compare identically-labeled Series objects

## Data Summarization and Visualization

# Instruction 2: Jupyter Notebook

In [5]:
!which python

/opt/anaconda3/envs/MyEnv/bin/python


In [6]:
!which jupyter

/opt/anaconda3/envs/MyEnv/bin/jupyter


In [7]:
!which python3

/opt/anaconda3/envs/MyEnv/bin/python3


# Instruction 3: Github Repository, Handles, and Evaluation

https://github.com/Naren1610/IST5520GrpProj/

### IST5520GrpProj

<h4 align="left"> Ronald Adomako, adomakor412 </h3>
<h4 align="left"> Idris Dobbs, idobbs-2012 </h3>
<h4 align="left"> Narendra Chigili, Naren1610 </h3>
<h4 align="left"> Nikhil Srirama Sai, SaiNikhilPalaparthi </h3>